src/incoming-message-stream.ts

```diff
@@ -43,13 +43,13 @@ class IncomingMessageStream extends Transform {
     return this;
   }
 
-  processBufferedData(callback: () => void) {
+  processBufferedData(callback: (err?: Error) => void) {
     // The packet header is always 8 bytes of length.
     while (this.bl.length >= HEADER_LENGTH) {
       // Get the full packet length
       const length = this.bl.readUInt16BE(2);
 
-      if (this.bl.length >= length) {
+      if (this.bl.length >= length && length >= HEADER_LENGTH) {
```

Comment:
🤔 What happens if `this.bl.length` is smaller than `length` if not enough data has been received from the server side yet?

Reply:
Ok, so I double checked, and `this.bl.length` is always greater than `HEADER_LENGTH` (due to the condition in the `while` statement above).

I still think there's something not _quite_ right with this condition. Let's say the so-far received buffer length (`this.bl.length`) is `20` and `length` read from the header is `40`. In this case, we would also incorrectly call `callback` with an error, while before we correctly used `break` to exit the loop and wait for more data to come in.

I think what we need to do here is not change this `if` (or it's `else` block) at all, and instead put a condition right before it that checks that `length` is at least `HEAD_LENGTH`.

Something like this:

```js
if (length < HEADER_LENGTH) {
  return callback(new ConnectionError('Unable to process incoming packet'));
}
```